src/contrib/tf_op/tvm_dso_op_kernels.cc

```diff
@@ -0,0 +1,344 @@
+
+    static int device_id(OpKernelContext* context) {
+        auto device_base = context->device();
+        auto gpu_device_info = device_base->tensorflow_gpu_device_info();
+        return gpu_device_info->gpu_id;
+    }
+
+    static void make_shape_from_tensor(
+            const tensorflow::Tensor& shape_tensor,
+            tensorflow::TensorShape* output_shape) {
+        tensorflow::int64 num_dims = shape_tensor.NumElements();
+        const tensorflow::int64* flat =
+            shape_tensor.flat<tensorflow::int64>().data();
+        tensorflow::int64* dims = new tensorflow::int64[num_dims];
+        cudaMemcpy(dims, flat, sizeof(tensorflow::int64) * num_dims,
+           cudaMemcpyDeviceToHost);
+        tensorflow::TensorShapeUtils::MakeShape(
+            dims, num_dims, output_shape);
+        delete dims;
+    }
+};
+#endif
+
+
+template <typename DEVICE_TYPE, int NUM_INPUTS>
+class TVMDSOOp : public OpKernel {
+ private:
+  tvm::runtime::PackedFunc tvm_func;
+  std::string lib_path;
+  std::string func_name;
+
+  tensorflow::DataType output_dtype;
+
+  bool has_static_output_shape;
+  std::vector<tensorflow::int64> static_output_shape;
+
+  void initAttributes(OpKernelConstruction* context) {
+    context->GetAttr("lib_path", &lib_path);
+    context->GetAttr("func_name", &func_name);
+    context->GetAttr("output_dtype", &output_dtype);
+
+    context->GetAttr("has_static_output_shape", &has_static_output_shape);
+    context->GetAttr("static_output_shape", &static_output_shape);
+  }
+
+ public:
+  explicit TVMDSOOp(OpKernelConstruction* context) : OpKernel(context) {
+    // Get attr
+    initAttributes(context);
+
+    // Load TVM function from dynamic library
+    tvm::runtime::Module mod_dylib =
+        tvm::runtime::Module::LoadFromFile(lib_path);
+    tvm_func = mod_dylib.GetFunction(func_name);
+    CHECK(tvm_func != nullptr);
+  }
+
+  void Compute(tensorflow::OpKernelContext* context) override {
+    DLTensor args[NUM_INPUTS + 1];
+    TensorAsBuf buf_info[NUM_INPUTS];
+    ShapeContainer shapes[NUM_INPUTS];
+
+    tensorflow::Status status;
+    int device_id = TVMDSOOpTrait<DEVICE_TYPE>::device_id(context);
+    int device_type = TVMDSOOpTrait<DEVICE_TYPE>::device_type;
+
+    DLContext dl_ctx = { DLDeviceType(device_type), device_id };
+
+    // Get output shape
+    tensorflow::TensorShape output_shape;
+    auto& output_shape_tensor = context->input(NUM_INPUTS);
+    if (has_static_output_shape) {
+      // use static output shape
+      const tensorflow::int64* dims = static_output_shape.data();
+      tensorflow::TensorShapeUtils::MakeShape(
+          dims, static_output_shape.size(), &output_shape);
+    } else if (output_shape_tensor.dims() == 1) {
+      // use shape tensor values as output shape
+      TVMDSOOpTrait<DEVICE_TYPE>::make_shape_from_tensor(
+          output_shape_tensor, &output_shape);
+    } else {
+      // use input tensor shape by default
+      output_shape = context->input(0).shape();
```

Comment:
Out of curiosity, how does this work? What happens if output is larger than this?

Reply:
Yeah it is not a safe code path. It would be better that output shape must be provided at the beginning.
```python
def func(self, name, output_dtype=None, output_shape=None):
     """Get tvm op function wrapped as TensorFlow tensor to tensor function```
```